### PR TITLE
BFD-2992: Prune old Platinum/App AMIs

### PR DESCRIPTION
**Pull request by keithdadkins**
**JIRA Ticket:**
[BFD-2992](https://jira.cms.gov/browse/BFD-2992)
[BFD-2993](https://jira.cms.gov/browse/BFD-2993)

**User Story or Bug Summary:**

Implement AMI retention policy
---

### What Does This PR Do?

- Adds a new module in mgmt for managing backups (trying to future-proof here)
- Adds retention policies for Platinum and App amis
- Adds a lambda that finds and prunes old Platinum/Application AMI's that fall outside of their respective retention policy
- Adds an event trigger and cron job to fire the prune job each night
- Updates our Launch Templates to keep its $DefaultVersion updated (see below)
- Removes the manual pruning script as it's no longer needed

### Pruning Overview
The lambda works by building a pool of candidate AMI's based on a set of resource Filters defined near the top of the
script. It then iterates through the pool and removes candidates still in their retention period. In addition, we **always** retain:

- Any AMI or EC2 Snapshot tagged with any `KEEP_TAGS` ('do-not-delete', 'keep', etc)
- Any EC2 Snapshot associated with an existing EC2 Volume
- Any AMI referenced by the $Latest or $Default version of an existing Launch Template

Remaining candidates are then deregistered and their associated snapshots deleted.

The script's complexity comes from the fact that some of our apps use launch templates and some do not. Launch templates are versioned and it's possible to delete a previous version. Launch Templates also identify two special versions: Latest and Default. We always retain both. I also included updates to our Launch Template configs to keep the DefaultVersion updated as we go, ie. so it points to the latest version instead of default version number 1 (the default version is what's used if you manually create an instance from the template without specifying a version).

### Terraform

Minified/redacted terraform plan
```sh
$ terraform plan
Terraform will perform the following actions:
  # aws_cloudwatch_event_rule.ami_pruner_schedule will be created
  + resource "aws_cloudwatch_event_rule" "ami_pruner_schedule" {
      ...
      + description         = "Cron schedule for running the AMI pruner"
      + schedule_expression = "cron(0 2 * * ? *)"
      ... 
    }

  # aws_cloudwatch_event_target.ami_pruner_schedule will be created
  ...

  # aws_cloudwatch_log_group.ami_pruner will be created
  + resource "aws_cloudwatch_log_group" "ami_pruner" {
      ...
      + name              = "/aws/lambda/bfd-mgmt-backups-ami-pruner"
      + retention_in_days = 30
      ...
    }

  # aws_iam_role.ami_pruner will be created
  ...

  # aws_iam_role_policy.ami_pruner will be created
  + resource "aws_iam_role_policy" "ami_pruner" {
      ...
              + Statement = [
                  + {
                      + Action   = [
                          + "ec2:DescribeImages",
                          + "ec2:DeregisterImage",
                          + "ec2:DescribeSnapshots",
                          + "ec2:DeleteSnapshot",
                          + "ec2:DescribeInstances",
                          + "ec2:DescribeVolumes",
                          + "ec2:DescribeTags",
                        ]
                      + Effect   = "Allow"
                      + Resource = [
                          + "arn:aws:ec2:us-east-1:$ACCTNUM:*",
                        ]
                      + Sid      = "AllowAmiPruning"
                    },
                  + {
                      + Action   = [
                          + "ssm:GetParameter",
                          + "ssm:GetParameters",
                          + "ssm:GetParameterByPath",
                        ]
                      + Effect   = "Allow"
                      + Resource = [
                          + "arn:aws:ssm:us-east-1:$ACCTNUM:parameter/bfd/mgmt/common/sensitive/backups/ami/*",
                        ]
                      + Sid      = "AllowRetentionPolicyLookup"
                    },
                  + {
                      + Action   = [
                          + "logs:CreateLogGroup",
                        ]
                      + Effect   = "Allow"
                      + Resource = [
                          + "arn:aws:logs:us-east-1:$ACCTNUM:*",
                        ]
                      + Sid      = "AllowLogGroupCreation"
                    },
                  + {
                      + Action   = [
                          + "logs:CreateLogGroup",
                          + "logs:CreateLogStream",
                          + "logs:PutLogEvents",
                        ]
                      + Effect   = "Allow"
                      + Resource = [
                          + "arn:aws:logs:us-east-1:$ACCTNUM:log-group:/aws/lambda/bfd-mgmt-backups-ami-pruner:*",
                        ]
                      + Sid      = "AllowLogging"
                    },
                  + {
                      + Action   = [
                          + "kms:Decrypt",
                        ]
                      + Effect   = "Allow"
                      + Resource = [
                          + "arn:aws:kms:us-east-1:$ACCTNUM:key/$KMSKEYID",
                        ]
                      + Sid      = "AllowKmsDecryption"
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + role        = (known after apply)
    }

  # aws_lambda_function.ami_pruner will be created
  + resource "aws_lambda_function" "ami_pruner" {
      ...
      + memory_size                    = 128
      + runtime                        = "python3.11"
      + timeout                        = 900
      ...
    }

  # aws_lambda_permission.ami_pruner_schedule will be created
  ...
Plan: 7 to add, 0 to change, 0 to destroy.
```
### Pruner

Screenshot from a dry run made on 10/24
ami-prune-dry-run

### What Should Reviewers Watch For?

- As of 10/24, mgmt's base_config/values eyaml is up-to-date with master (with the addition of the ami policies). This branch will need updating if this file is modified.
- While I'm confident in the pruning script, I'm less so with regards to the cron trigger. If something sticks out, please let me know.

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: ops/terraform/env/mgmt/backups/lambda-src/ami_pruner/ami_pruner.py [hunk 1]
```diff
@@ -252,7 +253,6 @@ def prune_candidates(candidates) -> (int, int):
     deregistered = []
     deleted = []
 
-    # scream, aim, and fire
```
Comment (brandoncruz3) [inline ops/terraform/env/mgmt/backups/lambda-src/ami_pruner/ami_pruner.py]: 🤣

**Review by brandoncruz3 — APPROVED** [review]
🚀

**Comment by keithdadkins** [comment]
@brandoncruz3 Had to rebase in and force push changes related to the new multi-region keys if you don't mind another look.

**Review by brandoncruz3 — APPROVED** [review]

**Review by mjburling — APPROVED** [review]
This looks great, in spite of the complexities of non-ASG managed EC2 instances, etc. Looking forward to scheduled runs in the near future!

**Pull request merged**